dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java

```diff
@@ -0,0 +1,73 @@
+package com.dat3m.dartagnan.program.processing;
+
+import com.dat3m.dartagnan.program.Program;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.sosy_lab.common.configuration.Configuration;
+import org.sosy_lab.common.configuration.InvalidConfigurationException;
+import org.sosy_lab.common.configuration.Option;
+import org.sosy_lab.common.configuration.Options;
+
+import static com.dat3m.dartagnan.configuration.OptionNames.*;
+
+@Options
+public class ProcessingManager implements ProgramProcessor {
+
+    private final static Logger logger = LogManager.getLogger(ProcessingManager.class);
+
+    private final Configuration config;
+
+    // =========================== Configurables ===========================
+
+    @Option(name = REDUCE_SYMMETRY,
+            description = "Reduces the symmetry of the program (unsound in general).",
+            secure = true)
+    private boolean reduceSymmetry = false;
+
+	@Option(name= ATOMIC_BLOCKS_AS_LOCKS,
+			description="Transforms atomic blocks by adding global locks.",
+			secure=true)
+		private boolean atomicBlocksAsLocks = false;
+
+	@Option(name= CONSTANT_PROPAGATION,
+			description="Perform constant propagation.",
+			secure=true)
+		private boolean constantPropagation = false;
+
+    // ======================================================================
+
+    private ProcessingManager(Configuration config) throws InvalidConfigurationException {
+        this.config = config;
+        config.inject(this);
+    }
+
+    public static ProcessingManager fromConfig(Configuration config) throws InvalidConfigurationException {
+        return new ProcessingManager(config);
+    }
+
+    // ==================================================
+
+    public void run(Program program){
+        try {
+            DeadCodeElimination.fromConfig(config).run(program);
+            BranchReordering.fromConfig(config).run(program);
+            Simplifier.fromConfig(config).run(program);
+            LoopUnrolling.fromConfig(config).run(program);
+            if(constantPropagation) {
+              ConstantPropagation.fromConfig(config).run(program);            	
+            }
+            Compilation.fromConfig(config).run(program);
+            if(atomicBlocksAsLocks) {
```

Comment:
I think I'm moved this at some point. If the order of a pass is important, we should add some precondition to it (as we do for e.g. compilation or constant propagation)

Reply:
The Precondition checks a processing pass should only be related to the processing step itself, that is, it should only check for properties that have to hold for the pass to work. If there is some other reason that forces us to order processing steps (i.e. the steps are interchangable in principle but the combined result is not what we want) then these checks should be inside the `ProcessingManager` either via `assert` or `Verify.verify(...)`.

As an example: `ConstantPropagation` assumes no loops, so it can have a Precondition that checks that the program is unrolled. But whether the program is compiled or not does not matter much, it should work in both cases. However, we have reasons of why we do not want to do it after compilation, so the `ProcessingManager` should have an assertion that checks that the program is not compiled before calling the `ConstantPropagation` code.
In the case of `AtomicBlocksAsLocks`, the code comments say that it can be used anytime in principal, so it should not have any Precondition, yet we still might want to run it before some other passes.

Reply:
Of course, if you rely on the fact that no exclusive events are present, then it should be a precondition on the pass.
But `AtomicBlocksAsLocks` does not have this assumption, so we shouldn't add any preconditions to it.
And in fact, it might be correct to execute at any time we want. The comment I added was just because we used to do this as a first step back then and that has changed. It doesn't mean its wrong though.

Reply:
From the code comments, it seems that this pass is actually non-functional right now. I haven't tested it, but we might be better off removing this for now especially because I think that the parsing code still performs this transformation.

Reply:
Since the option is set off and the code is completely in a separate class, I'm fine with keeping the code and fixing it in a different PR.

Reply:
My suggestion is to remove it only from the `ProcessingManager` and not completely. But I guess we have no users that would accidentally enable this option so it doesn't really matter I guess. Then let's keep it and add a TODO/FIXME

Reply:
`AtomicBlocksAsLocks` already has both

Reply:
You are right.